### PR TITLE
 DEPR: deprecate fastpath keyword in Index constructors 

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
From all that looking at Index constructors ... 

Apparently it was even hardly used internally for the Index classes (except a little bit in RangeIndex).  

Part of https://github.com/pandas-dev/pandas/issues/20110

### CODE REVIEW

File: pandas/core/indexes/period.py [hunk 1]
```diff
@@ -306,7 +306,7 @@ def __contains__(self, key):
 
     @cache_readonly
     def _int64index(self):
-        return Int64Index(self.asi8, name=self.name, fastpath=True)
+        return Int64Index._simple_new(self.asi8, name=self.name)
```
Comment (jbrockmendel) [inline pandas/core/indexes/period.py]: What is the performance implication of using _simple_new vs just Int64Index(...)?  If its sufficiently small, we should be using the public constructors

File: pandas/core/indexes/range.py [hunk 2]
```diff
@@ -174,7 +179,7 @@ def _data(self):
 
     @cache_readonly
     def _int64index(self):
-        return Int64Index(self._data, name=self.name, fastpath=True)
+        return Int64Index._simple_new(self._data, name=self.name)
```
Comment (jbrockmendel) [inline pandas/core/indexes/range.py]: shallow_copy?  ditto below

**Comment by jbrockmendel** [comment]
General +1, couple of comments.  Is there a tracker issue for deprecations-to-eventually-remove?

**Comment by jreback** [comment]
yes
search for the label of Deprecation then sort by most comments

File: pandas/core/indexes/range.py (hunk 2 above)
Comment (jorisvandenbossche) [reply pandas/core/indexes/range.py]: And actually, `_shallow_copy` has not the ability to specify (override) the `name` (which is not really following how `_shallow_copy` works for other Index classes), so can also not use it there.

Might be a bit buggy in `_shallow_copy` (it is also not used at all anywhere in the RangeIndex implementation, although maybe in inherited methods ..). 
But I want to keep it here just about deprecating the fastpath keyword, so let's leave that for another issue.

**Comment by jorisvandenbossche** [comment]
Added it to the list in https://github.com/pandas-dev/pandas/issues/6581

File: pandas/core/indexes/period.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/indexes/period.py]: @jbrockmendel this leads a bit further, but given that the public constructors (`__init__`) of Index classes are very complex (something we will not be able to change I think), I think that we need to see `_simple_new` as kind of a "internal public" method within the pandas code base.

**Comment by jorisvandenbossche** [comment]
@jbrockmendel Thanks for the review!

File: pandas/core/indexes/range.py (hunk 2 above)
Comment (jreback) [reply pandas/core/indexes/range.py]: why are u changing these constructors at all?

**Review by jreback — CHANGES_REQUESTED** [review]
i would prefer simply to use the actual constructor rather than _simple_new if we can

File: pandas/core/indexes/range.py (hunk 2 above)
Comment (jorisvandenbossche) [reply pandas/core/indexes/range.py]: I am not changing anything. `fastpath=True` means exactly the `_simple_new` what I replaced it with. 
This PR does *not* try to do a clean-up of usage of `_simple_new` by replacing it with the default constructor, it only deprecates the `fastpath` argument.

**Comment by jreback** [comment]
@jorisvandenbossche can you update

**Comment by jorisvandenbossche** [comment]
What do I need to update?

**Comment by jreback** [comment]
> why are u changing these constructors at all?

comment above

**Comment by jorisvandenbossche** [comment]
Answer above: https://github.com/pandas-dev/pandas/pull/23110#discussion_r225005757

**Review by jreback — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
thanks!

**Pull request merged**